src/GafferML/Tensor.cpp

```diff
@@ -0,0 +1,405 @@
+			break;
+		case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT32 :
+			functor( value.GetTensorData<int32_t>() );
+			break;
+		case ONNX_TENSOR_ELEMENT_DATA_TYPE_UINT64 :
+			functor( value.GetTensorData<uint64_t>() );
+			break;
+		case ONNX_TENSOR_ELEMENT_DATA_TYPE_INT64 :
+			functor( value.GetTensorData<int64_t>() );
+			break;
+		default :
+			throw IECore::Exception( fmt::format( "Unsupported element type {}", elementType ) );
+	}
+}
+
+DataPtr dataFromValue( const Ort::Value &value )
+{
+	DataPtr result;
+	dispatchTensorData(
+		value,
+		[&] ( const auto *data ) {
+
+			using ElementType = remove_const_t<remove_pointer_t<decltype( data )>>;
+			using DataType = TypedData<vector<ElementType>>;
+			using PtrType = typename DataType::Ptr;
+
+			PtrType d = new DataType;
+			const size_t count = value.GetTensorTypeAndShapeInfo().GetElementCount();
+			d->writable().insert( d->writable().end(), data, data + count );
+			result = d;
+
+		}
+	);
+	return result;
+}
+
+} // namespace
+
+IE_CORE_DEFINEOBJECTTYPEDESCRIPTION( Tensor );
+
+Tensor::State::State( Ort::Value &&value, IECore::ConstDataPtr data )
+	:	value( std::move( value ) ), data( data )
+{
+	if( value && !value.IsTensor() )
+	{
+		/// \todo Maybe we'll loosen this restriction at some point,
+		/// or maybe we'll create wrappers for the other Value types.
+		/// For the moment we just want to know if something unexpected
+		/// happens.
+		throw IECore::Exception( "Ort::Value is not a tensor" );
+	}
+}
+
+Tensor::Tensor()
+	:	m_state( new State( Ort::Value( nullptr ) ) )
+{
+}
+
+Tensor::Tensor( Ort::Value &&value )
+	:	m_state( new State( std::move( value ) ) )
+{
+}
+
+Tensor::Tensor( const IECore::ConstDataPtr &data, std::vector<int64_t> shape )
+{
+	IECore::dispatch(
+		data.get(),
+		[&] ( auto typedData ) -> void {
+
+			using DataType = remove_const_t<remove_pointer_t<decltype( typedData )>>;
+			using BaseType = typename DataType::BaseType;
+
+			if( !shape.size() )
+			{
+				// Automatically infer shape from type.
+				if constexpr( TypeTraits::IsVectorTypedData<DataType>::value )
+				{
+					shape.push_back( typedData->readable().size() );
+					using ShapeT = ShapeTraits<typename DataType::ValueType::value_type>;
+					shape.insert( shape.end(), begin( ShapeT::shape ), end( ShapeT::shape ) );
+				}
+				else
+				{
+					using ShapeT = ShapeTraits<typename DataType::ValueType>;
+					shape.insert( shape.end(), begin( ShapeT::shape ), end( ShapeT::shape ) );
+				}
+			}
+
+			if constexpr( std::is_same_v<DataType, BoolVectorData> )
+			{
+				// Special case for the vector of bool fiasco.
+				Ort::AllocatorWithDefaultOptions allocator;
+				Ort::Value value = Ort::Value::CreateTensor(
+					allocator, shape.data(), shape.size(), ONNX_TENSOR_ELEMENT_DATA_TYPE_BOOL
+				);
+				std::copy( typedData->readable().begin(), typedData->readable().end(), value.GetTensorMutableData<bool>() );
+				m_state = new State{ std::move( value ), nullptr };
```

Comment:
Because the `Ort::Value` owns its own copy of the data in this case. We only keep the DataPtr when the `Ort::Value` is referencing the data that we own within that `DataPtr`.

Reply:
Makes sense